orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -403,114 +405,150 @@ private static boolean isRequired(TypeDescription orcType) {
+    }
 
-        return getIcebergType(icebergID, name, Types.StructType.of(fields), isRequired);
-      }
-      case LIST: {
-        TypeDescription elementType = orcType.getChildren().get(0);
-        Types.NestedField element = convertOrcToIceberg(elementType, "element", nextID);
-
-        Types.ListType listTypeWithElem = isRequired(elementType) ?
-            Types.ListType.ofRequired(element.fieldId(), element.type()) :
-            Types.ListType.ofOptional(element.fieldId(), element.type());
-        return isRequired ?
-            Types.NestedField.required(icebergID, name, listTypeWithElem) :
-            Types.NestedField.optional(icebergID, name, listTypeWithElem);
+    @Override
+    public Optional<Types.NestedField> record(TypeDescription record, List<String> names,
+                                              List<Optional<Types.NestedField>> fields) {
+      boolean isRequired = isRequired(record);
+      Optional<Integer> icebergIdOpt = icebergID(record);
+      if (!icebergIdOpt.isPresent() || fields.size() == 0) {
+        return Optional.empty();
       }
-      case MAP: {
-        TypeDescription keyType = orcType.getChildren().get(0);
-        Types.NestedField key = convertOrcToIceberg(keyType, "key", nextID);
-        TypeDescription valueType = orcType.getChildren().get(1);
-        Types.NestedField value = convertOrcToIceberg(valueType, "value", nextID);
 
-        Types.MapType mapTypeWithKV = isRequired(valueType) ?
-            Types.MapType.ofRequired(key.fieldId(), value.fieldId(), key.type(), value.type()) :
-            Types.MapType.ofOptional(key.fieldId(), value.fieldId(), key.type(), value.type());
+      Types.StructType structType = Types.StructType.of(
+          fields.stream().filter(Optional::isPresent).map(Optional::get).collect(Collectors.toList()));
+      return Optional.of(getIcebergType(icebergIdOpt.get(), icebergToOrcMapping.get(icebergIdOpt.get()).name(),
+          structType, isRequired));
+    }
+
+    @Override
+    public Optional<Types.NestedField> list(TypeDescription array,
+                                            Optional<Types.NestedField> element) {
+      boolean isRequired = isRequired(array);
+      Optional<Integer> icebergIdOpt = icebergID(array);
 
-        return getIcebergType(icebergID, name, mapTypeWithKV, isRequired);
+      if (!icebergIdOpt.isPresent() || !element.isPresent()) {
+        return Optional.empty();
       }
-      default:
-        // We don't have an answer for union types.
-        throw new IllegalArgumentException("Can't handle " + orcType);
+
+      Types.NestedField foundElement = element.get();
+      Types.ListType listTypeWithElem = isRequired(array.getChildren().get(0)) ?
+          Types.ListType.ofRequired(foundElement.fieldId(), foundElement.type()) :
+          Types.ListType.ofOptional(foundElement.fieldId(), foundElement.type());
+      return Optional.of(getIcebergType(icebergIdOpt.get(),
```

Comment:
I don't think it is necessary to call `getIcebergType` (which returns a `NestedField`?).

Iceberg added `Types.NestedField.of()` that accepts a boolean `isOptional` so you don't have to call just `required` or `optional`. That could be used in place of `getIcebergType`.